peer/ping_manager.go

```diff
@@ -143,53 +175,64 @@ func (m *PingManager) pingHandler() {
 
 			// Set up our bookkeeping for the new Ping.
 			if err := m.setPingState(pongSize); err != nil {
-				m.cfg.OnPongFailure(err)
-
+				// This is an internal error related to timer
+				// reset. Pass it to OnPongFailure as it's
+				// critical. Current and last RTT are not
+				// directly applicable here.
+				m.cfg.OnPongFailure(err, 0, 0)
 				return
 			}
 
 			m.cfg.SendPing(ping)
 
 		case <-m.pingTimeout.C:
-			m.resetPingState()
-
-			e := errors.New("timeout while waiting for " +
-				"pong response",
+			timeWaited := m.getTimeWaitedOrDefault(
+				m.cfg.TimeoutDuration,
 			)
+			lastRTT := m.getLastRTT()
 
-			m.cfg.OnPongFailure(e)
+			m.cfg.OnPongFailure(
+				errors.New("timeout while waiting for "+
+					"pong response"),
+				timeWaited, lastRTT,
+			)
 
-			return
+			m.resetPingState()
```

Comment:
We previously returned here. Maybe we need to signal from `OnPongFailure` whether we actually disconnected. And if we did, we should still return here? Otherwise the ping handler will continue running, even if we disconnect.

Reply:
Hmm, good point. I think to resolve this, we can pass in the actual peer context into the ping manager. Then this way, the ping manager will _always_ shutdown when/if the peer disconnects.

Reply:
Pushed this up as a new commit.